crates/node/src/sync/manager.rs

```diff
@@ -755,9 +755,126 @@ impl SyncManager {
+                    %context_id,
+                    %chosen_peer,
+                    our_root_hash = %context.root_hash,
+                    peer_root_hash = %peer_root_hash,
+                    our_heads_count = context.dag_heads.len(),
+                    peer_heads_count = peer_dag_heads.len(),
+                    "Root hash mismatch with peer, triggering DAG catchup"
+                );
+
+                let our_heads_set: std::collections::HashSet<_> =
+                    context.dag_heads.iter().collect();
+                let missing_heads: Vec<_> = peer_dag_heads
+                    .iter()
+                    .filter(|h| !our_heads_set.contains(h))
+                    .cloned()
+                    .collect();
+
+                if !missing_heads.is_empty() {
+                    info!(
+                        %context_id,
+                        missing_count = missing_heads.len(),
+                        "Peer has DAG heads we don't have, requesting them"
+                    );
+
+                    let result = self
+                        .request_dag_heads_and_sync(context_id, chosen_peer, our_identity, stream)
+                        .await?;
+
+                    // If peer had no data or unexpected response, return error to try next peer
+                    if matches!(result, SyncProtocol::None) {
+                        bail!("Peer has no data or unexpected response for this context, will try next peer");
+                    }
+
+                    return Ok(Some(result));
+                } else {
+                    // Same heads but different root hash - may have deltas that haven't been applied yet
+                    warn!(
+                        %context_id,
+                        "Same DAG heads but different root hash - requesting full sync"
+                    );
+
+                    let result = self
+                        .request_dag_heads_and_sync(context_id, chosen_peer, our_identity, stream)
+                        .await?;
+
+                    // If peer had no data or unexpected response, return error to try next peer
+                    if matches!(result, SyncProtocol::None) {
+                        bail!("Peer has no data or unexpected response for this context, will try next peer");
+                    }
+
+                    return Ok(Some(result));
+                }
+            } else {
+                debug!(
+                    %context_id,
+                    root_hash = %context.root_hash,
+                    "Root hash matches peer, node is truly in sync"
+                );
+            }
+        }
+
         Ok(None)
     }
 
+    /// Query peer for their DAG state (root_hash and dag_heads) without triggering full sync
+    async fn query_peer_dag_state(
+        &self,
+        context_id: ContextId,
+        our_identity: PublicKey,
+        stream: &mut Stream,
+    ) -> eyre::Result<Option<(calimero_primitives::hash::Hash, Vec<[u8; 32]>)>> {
```

Comment:
Please add a full doc comment here. It's especially important to describe the return value.